python/cuda_parallel/benchmarks/bench_cuda_parallel.py

```diff
@@ -0,0 +1,169 @@
+
+    cp.cuda.runtime.deviceSynchronize()
+
+
+def merge_sort_iterator(size, build_only):
+    keys_dt = cp.int32
+    vals_dt = cp.int64
+    keys = iterators.CountingIterator(np.int32(0))
+    vals = iterators.CountingIterator(np.int64(0))
+    res_keys = cp.empty(size, dtype=keys_dt)
+    res_vals = cp.empty(size, dtype=vals_dt)
+
+    def my_cmp(a: np.int32, b: np.int32) -> np.int32:
+        return np.int32(a < b)
+
+    alg = algorithms.merge_sort(keys, vals, res_keys, res_vals, my_cmp)
+    temp_bytes = alg(None, keys, vals, res_keys, res_vals, size)
+    scratch = cp.empty(temp_bytes, dtype=cp.uint8)
+
+    if not build_only:
+        alg(scratch, keys, vals, res_keys, res_vals, size)
+
+    cp.cuda.runtime.deviceSynchronize()
+
+
+@pytest.fixture(params=[True, False])
+def build_only(request):
+    return request.param
+
+
+@pytest.fixture(params=[10_000, 100_000, 1_000_000])
+def size(request):
+    return request.param
+
+
+def bench_compile_reduce_pointer(benchmark):
+    def setup():
+        # This function is called once before the benchmark runs
+        # to set up the environment.
+        algorithms.reduce_into.cache_clear()
+
+    benchmark.pedantic(
+        reduce_pointer,
+        kwargs={"size": 10, "build_only": True},
+        rounds=3,
+        iterations=1,
+        setup=setup,
+    )
+
+
+def bench_compile_reduce_iterator(benchmark):
+    def setup():
+        # This function is called once before the benchmark runs
+        # to set up the environment.
+        algorithms.reduce_into.cache_clear()
+
+    benchmark.pedantic(
+        reduce_iterator,
+        kwargs={"size": 10, "build_only": True},
+        rounds=3,
+        iterations=1,
+        setup=setup,
+    )
+
+
+def bench_compile_merge_sort_pointer(benchmark):
+    def setup():
+        # This function is called once before the benchmark runs
+        # to set up the environment.
+        algorithms.merge_sort.cache_clear()
+
+    benchmark.pedantic(
+        merge_sort_pointer,
+        kwargs={"size": 10, "build_only": True},
+        rounds=3,
+        iterations=1,
+        setup=setup,
+    )
+
+
+def bench_compile_merge_sort_iterator(benchmark):
+    def setup():
+        # This function is called once before the benchmark runs
+        # to set up the environment.
+        algorithms.merge_sort.cache_clear()
+
+    benchmark.pedantic(
+        merge_sort_iterator,
+        kwargs={"size": 10, "build_only": True},
+        rounds=3,
+        iterations=1,
+        setup=setup,
+    )
+
+
+def bench_reduce_pointer(benchmark, size):
+    benchmark(reduce_pointer, size, build_only=False)
+
+
+def bench_reduce_iterator(benchmark, size):
+    benchmark(reduce_iterator, size, build_only=False)
+
+
+def bench_merge_sort_pointer(benchmark, size):
+    benchmark(merge_sort_pointer, size, build_only=False)
+
+
+def bench_merge_sort_iterator(benchmark, size):
+    benchmark(merge_sort_iterator, size, build_only=False)
```

Comment:
I am out of my league here but are we sure we want to be maintaining a monolithic file for all the benchmarks? Python benchmarking infra looks simplistic rn as it's maturing but maybe it would be better if we start off by splitting them in separate files. Sth like:

```
benchmarks
   |---- pybench_util.py (for any common utility code)
   |---- reduce.py (or reduce_bench.py)
   |---- merge_sort.py
```

If not, maybe just renaming the monolithic file to something more specific atm.

Reply:
- I split up `bench_cuda_parallel.py` into `bench_reduce.py` and `bench_merge_sort.py`
- I put all common functionality in a [`conftest.py`](https://docs.pytest.org/en/stable/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files)